switchio/loops/proactor.py

```diff
@@ -0,0 +1,330 @@
+
+                callbacks = self.callbacks.get(cid, False)
+                if callbacks and consumed:
+                    cbs = callbacks.get(evname, ())
+                    self.log.debug(
+                        "consumer '{}' has callback {} registered for ev {}"
+                        .format(cid, cbs, evname)
+                    )
+                    # look up the client's callback chain and run
+                    # e -> handler -> cb1, cb2, ... cbN
+                    # XXX assign ret on each interation in an attempt to avoid
+                    # python's dynamic scope lookup
+                    for cb, ret in zip(cbs, itertools.repeat(ret)):
+                        try:
+                            cb(*ret)
+                        except Exception:
+                            self.log.exception(
+                                "Failed to execute callback {} for event "
+                                "with uid {}".format(cb, uid)
+                            )
+                if model:
+                    # signal any awaiting futures
+                    fut = model._futures.get(evname, None)
+                    if fut:
+                        fut.set_result(e)
+                        # resume waiting coroutines...
+                        # seriously guys, this is literally so stupid
+                        # and confusing
+                        await just_yield()
+
+                coroutines = self.coroutines.get(cid, False)
+                if coroutines and consumed:
+                    coros = coroutines.get(evname, ())
+                    self.log.debug(
+                        "app '{}' has coroutines {} registered for ev {}"
+                        .format(cid, coros, evname)
+                    )
+                    # look up and schedule assigned coroutines
+                    # e -> handler -> coro1, coro2, ... coroN
+                    for coro in coros:
+                        task = asyncio.ensure_future(coro(*ret), loop=loop)
+                        self._tasks.append(task)
+
+                        # unblock `session.vars` waiters
+                        if model in self._sess2waiters:
+                            for var, evs in self._sess2waiters[model].items():
+                                if model.vars.get(var):
+                                    [event.set() for event in evs]
+
+            # exception raised by handler/chain on purpose?
+            except utils.ESLError:
+                consumed = True
+                self.log.warning("Caught ESL error for event '{}':\n{}"
+                                 .format(evname, traceback.format_exc()))
+            except Exception:
+                self.log.exception(
+                    "Failed to process event {} with uid {}"
+                    .format(evname, uid)
+                )
+            return consumed
+        else:
+            self.log.error("Unknown event '{}'".format(evname))
```

Comment:
You should turn on flake8's cyclomatic complexity measurements 😉

Reply:
Ok i'll make an issue to investigate.